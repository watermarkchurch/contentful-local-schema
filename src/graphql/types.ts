import type { Resolver } from '@apollo/client'
import type { Asset as ContentfulAsset } from '../contentful/types'
import { GraphQLFloat, GraphQLInt, GraphQLInterfaceType, GraphQLList, GraphQLNonNull, GraphQLObjectType, GraphQLScalarType, GraphQLString } from 'graphql'
import inflection from 'inflection'

export const GraphQLNever = new GraphQLScalarType({
  name: 'Never',
  serialize: () => null
})

export type Namespace = ReturnType<typeof namespace>

export function namespace(ns?: string) {
  const namespaceName = ns ? inflection.titleize(ns) : undefined

  const GraphQLLocation  = new GraphQLObjectType({
    name: toType('Location'),
    fields: {
      lat: { type: GraphQLFloat },
      lon: { type: GraphQLFloat }
    }
  })

  const Sys = new GraphQLObjectType({
    name: toType('Sys'),
    fields: {
      id: { type: GraphQLString },
      spaceId: { type: GraphQLString },
      environmentId: { type: GraphQLString },
    }
  })
  const ContentfulTag = new GraphQLObjectType({
    name: toType('ContentfulTag'),
    fields: {
      id: { type: new GraphQLNonNull(GraphQLString) },
      name: { type: new GraphQLNonNull(GraphQLString) },
    }
  })

  const ContentfulMetadata = new GraphQLObjectType({
    name: toType('ContentfulMetadata'),
    fields: {
      tags: { type: new GraphQLNonNull(new GraphQLList(ContentfulTag)) }
    }
  })

  const Entry = new GraphQLInterfaceType({
    name: toType('Entry'),
    fields: {
      sys: { type: Sys },
      contentfulMetadata: { type: ContentfulMetadata }
    }
  })

  const EntryCollection = new GraphQLObjectType({
    name: toType('EntryCollection'),
    fields: {
      skip: { type: new GraphQLNonNull(GraphQLInt) },
      limit: { type: new GraphQLNonNull(GraphQLInt) },
      total: { type: new GraphQLNonNull(GraphQLInt) },
      items: { type: new GraphQLNonNull(new GraphQLList(Entry)) }
    }
  })

  const Asset = new GraphQLObjectType({
    name: toType('Asset'),
    fields: {
      sys: { type: Sys },
      contentfulMetadata: { type: ContentfulMetadata },
      // linkedFrom: AssetLinkingCollections
      title: { type: GraphQLString },
      description: { type: GraphQLString },
      contentType: { type: GraphQLString },
      fileName: { type: GraphQLString },
      url: { type: GraphQLString },
      size: { type: GraphQLInt },
      width: { type: GraphQLInt },
      height: { type: GraphQLInt },
    }
  })

  const AssetCollection = new GraphQLObjectType({
    name: toType('AssetCollection'),
    fields: {
      skip: { type: new GraphQLNonNull(GraphQLInt) },
      limit: { type: new GraphQLNonNull(GraphQLInt) },
      total: { type: new GraphQLNonNull(GraphQLInt) },
      items: { type: new GraphQLNonNull(new GraphQLList(Asset)) }
    }
  })

  return {
    _name: namespaceName,
    Asset,
    AssetCollection,
    ContentfulMetadata,
    ContentfulTag,
    Entry,
    EntryCollection,
    Sys,
    GraphQLLocation,
    toType
  }

  function toType(name: string): string {
    return namespacedTypeName(name, namespaceName)
  }
}

export function namespacedTypeName(name: string, namespace: string | undefined) {
  if (namespace && !name.startsWith(namespace + '_')) {
    name = `${namespace}_${name}`
  }
  return name
}

export function assetFieldResolver(namespace: string | undefined): { [field: string]: Resolver } {
  const SysTypeName = namespacedTypeName('Sys', namespace)
  return {
    sys: (asset: ContentfulAsset) => {
      return {
        __typename: SysTypeName,
        _id: asset.sys.id,
        ...asset.sys
      }
    },
    title: (asset: ContentfulAsset) => asset.fields.title || null,
    description: (asset: ContentfulAsset) => asset.fields.description || null,
    contentType: (asset: ContentfulAsset) => asset.fields.file?.contentType || null,
    fileName: (asset: ContentfulAsset) => asset.fields.file?.fileName || null,
    url: (asset: ContentfulAsset) => asset.fields.file?.url || null,
    size: (asset: ContentfulAsset) => asset.fields.file?.details?.size || null,
    width: (asset: ContentfulAsset) => asset.fields.file?.details?.image?.width || null,
    height: (asset: ContentfulAsset) => asset.fields.file?.details?.image?.height || null,
  }
}