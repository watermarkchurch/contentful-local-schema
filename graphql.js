
module.exports = require('./dist/graphql')

