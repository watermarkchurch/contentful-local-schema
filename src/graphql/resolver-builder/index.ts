import type {Resolvers} from '@apollo/client'

import { ContentfulDataSource } from '../../dataSource'
import { assetFieldResolver, namespacedTypeName } from '../types'
import { ContentType } from '../../contentful/types'
import ContentTypeResolverBuilder from './content-type-resolver-builder'
import QueryResolverBuilder from './query-resolver-builder'


interface IBaseOptions {
  namespace?: string
  queryNamespace?: string

  logger: { debug: Console['debug'] },
}

type IOptions = IBaseOptions & (
  {
    filename: string
  } | {
    contentTypes: any[]
  }
)

export default class ResolverBuilder {
  private readonly options: Readonly<IOptions>

  constructor(
    private readonly dataSource: ContentfulDataSource,
    options?: Partial<IOptions>
  ) {
    const opts: IOptions = Object.assign({
      filename: './contentful-schema.json',
      logger: console,
    } as IOptions, options)

    this.options = opts
  }

  public build(): Resolvers {
    let contentfulSchema: { contentTypes: ContentType[] }
    if ('contentTypes' in this.options) {
      contentfulSchema = { contentTypes: this.options.contentTypes }
    } else {
      contentfulSchema = require(this.options.filename)
    }
    const {namespace, queryNamespace} = this.options

    const resolvers: Resolvers = {
      ...contentfulSchema.contentTypes.reduce((types, ct) => {
        return {
          ...types,
          ...new ContentTypeResolverBuilder(this.dataSource, this.options, ct).build()
        }
      }, {})
    }

    const Query = {
      ...new QueryResolverBuilder(this.dataSource, this.options, 'Entry').build().Query,
      ...new QueryResolverBuilder(this.dataSource, this.options, 'Asset').build().Query,
      ...contentfulSchema.contentTypes.reduce((resolvers, ct) => {
        return {
          ...resolvers,
          ...new QueryResolverBuilder(this.dataSource, this.options, ct.sys.id).build().Query
        }
      }, {})
    }

    if (queryNamespace) {
      // we put the Query inside a namespaced type that we reference in the top level
      const QueryTypeName = namespacedTypeName('Query', queryNamespace)
      resolvers[QueryTypeName] = Query
      resolvers['Query'] = {}
      resolvers['Query'][queryNamespace] = () => {
        return {
          __typename: QueryTypeName,
          _id: QueryTypeName,
        }
      }
    } else {
      // we directly add our resolvers to the query
      resolvers['Query'] = Query
    }

    resolvers[namespacedTypeName('Asset', namespace)] = assetFieldResolver(namespace)

    return resolvers
  }

}
