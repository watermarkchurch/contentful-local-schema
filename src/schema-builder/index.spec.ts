import { printSchema } from "graphql"
import SchemaBuilder from "."

describe("SchemaBuilder", () => {
  it("writes schema", async () => {
    const instance = new SchemaBuilder({
      contentTypes: contentfulSchema.contentTypes
    })

    const schema = await instance.build()

    expect(printSchema(schema)).toMatchSnapshot()
  })
})

const contentfulSchema = {
  contentTypes: [
    {
      sys: {
        id: "section-block-text",
        type: "ContentType",
      },
      displayField: "internalTitle",
      name: "Section: Block Text",
      description: "Markdown free-text block",
      fields: [
        {
          id: "internalTitle",
          name: "Internal Title (Contentful Only)",
          type: "Symbol",
          localized: false,
          required: true,
          validations: [],
          disabled: false,
          omitted: true,
        },
        {
          id: "body",
          name: "Body",
          type: "Text",
          localized: false,
          required: true,
          validations: [],
          disabled: false,
          omitted: false,
        },
        {
          id: "bookmarkTitle",
          name: "Bookmark Title",
          type: "Symbol",
          localized: true,
          required: false,
          validations: [],
          disabled: false,
          omitted: false,
        },
        {
          id: "style",
          name: "Style",
          type: "Symbol",
          localized: false,
          required: false,
          validations: [
            {
              in: ["default", "narrow"],
            },
          ],
          disabled: false,
          omitted: false,
        },
      ],
    },
    {
      sys: {
        id: "page",
        type: "ContentType",
      },
      displayField: "internalTitle",
      name: "Page",
      description:
        "A page describes a collection of sections that correspondto a URL slug",
      fields: [
        {
          id: "title",
          name: "Title",
          type: "Symbol",
          localized: false,
          required: true,
          validations: [],
          disabled: false,
          omitted: false,
        },
        {
          id: "subpages",
          name: "Subpages",
          type: "Array",
          localized: false,
          required: false,
          validations: [],
          disabled: false,
          omitted: false,
          items: {
            type: "Link",
            validations: [
              {
                linkContentType: ["page"],
              },
            ],
            linkType: "Entry",
          },
        },
        {
          id: "meta",
          name: "Meta",
          type: "Link",
          localized: false,
          required: false,
          validations: [
            {
              linkContentType: ["pageMetadata"],
            },
          ],
          disabled: false,
          omitted: false,
          linkType: "Entry",
        },
        {
          id: "sections",
          name: "Sections",
          type: "Array",
          localized: false,
          required: false,
          validations: [],
          disabled: false,
          omitted: false,
          items: {
            type: "Link",
            validations: [
              {
                linkContentType: ["section-block-text", "section-carousel"],
              },
            ],
            linkType: "Entry",
          },
        },
        {
          id: "flags",
          name: "Flags",
          type: "Array",
          localized: false,
          required: false,
          validations: [],
          disabled: false,
          omitted: false,
          items: {
            type: "Symbol",
            validations: [
              {
                in: ["Not Shareable", "Set Campus Cookie"],
              },
            ],
          },
        },
      ],
    },
    {
      sys: {
        id: "section-carousel",
        type: "ContentType",
      },
      displayField: "internalTitle",
      name: "Section: Carousel",
      description:
        "Renders a carousel of images that automatically progresses from one to the next",
      fields: [
        {
          id: "images",
          name: "Images",
          type: "Array",
          localized: false,
          required: false,
          validations: [
            {
              size: {
                min: 1,
                max: null,
              },
            },
          ],
          disabled: false,
          omitted: false,
          items: {
            type: "Link",
            validations: [
              {
                linkMimetypeGroup: ["image"],
              },
              {
                assetFileSize: {
                  min: null,
                  max: 20971520,
                },
              },
            ],
            linkType: "Asset",
          },
        },
        {
          id: "slideInterval",
          name: "Slide Interval (Seconds)",
          type: "Number",
          localized: false,
          required: false,
          validations: [
            {
              range: {
                min: 1,
                max: null,
              },
            },
          ],
          disabled: false,
          omitted: false,
        },
      ],
    },
    {
      "sys": {
        "id": "pageMetadata",
        "type": "ContentType"
      },
      "displayField": "internalTitle",
      "name": "Page Metadata",
      "description": "Page Metadata contains SEO information about a page",
      "fields": [
        {
          "id": "canonicalLink",
          "name": "Canonical Link",
          "type": "Symbol",
          "localized": false,
          "required": false,
          "validations": [
            {
              "unique": true
            },
            {
              "regexp": {
                "pattern": "^(ftp|http|https):\\/\\/(\\w+:{0,1}\\w*@)?(\\S+)(:[0-9]+)?(\\/|\\/([\\w#!:.?+=&%@!\\-\\/]))?$"
              }
            }
          ],
          "disabled": false,
          "omitted": false
        },
        {
          "id": "metaKeywords",
          "name": "Meta Keywords",
          "type": "Text",
          "localized": false,
          "required": false,
          "validations": [],
          "disabled": false,
          "omitted": false
        },
        {
          "id": "metaFlag",
          "name": "Meta Flag",
          "type": "Array",
          "localized": false,
          "required": false,
          "validations": [],
          "disabled": false,
          "omitted": false,
          "items": {
            "type": "Symbol",
            "validations": [
              {
                "in": [
                  "no-follow",
                  "no-index"
                ]
              }
            ]
          }
        },
        {
          "id": "structuredData",
          "name": "Structured Data",
          "type": "Object",
          "localized": false,
          "required": false,
          "validations": [],
          "disabled": false,
          "omitted": false
        }
      ]
    }
  ]
}