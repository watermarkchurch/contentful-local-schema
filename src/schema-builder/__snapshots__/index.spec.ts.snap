// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SchemaBuilder writes schema 1`] = `
"type Query {
  asset(id: String): Asset
  assetCollection(skip: String, limit: String): AssetCollection
  sectionBlockText(id: String): SectionBlockText
  sectionBlockTextCollection(skip: String, limit: String): SectionBlockTextCollection
  page(id: String): Page
  pageCollection(skip: String, limit: String): PageCollection
  sectionCarousel(id: String): SectionCarousel
  sectionCarouselCollection(skip: String, limit: String): SectionCarouselCollection
  pageMetadata(id: String): PageMetadata
  pageMetadataCollection(skip: String, limit: String): PageMetadataCollection
}

type Asset {
  sys: Sys
  contentfulMetadata: ContentfulMetadata
  title: String
  description: String
  contentType: String
  fileName: String
  url: String
  size: Int
  width: Int
  height: Int
}

type Sys {
  id: String
  spaceId: String
  environmentId: String
}

type ContentfulMetadata {
  tags: [ContentfulTag]!
}

type ContentfulTag {
  id: String!
  name: String!
}

type AssetCollection {
  skip: Int!
  limit: Int!
  total: Int!
  items: [Asset]!
}

type SectionBlockText implements Entry {
  internalTitle: Never
  body: String!
  bookmarkTitle: String
  style: SectionBlockTextStyle
}

interface Entry {
  sys: Sys
  contentfulMetadata: ContentfulMetadata
}

scalar Never

enum SectionBlockTextStyle {
  default
  narrow
}

type SectionBlockTextCollection {
  skip: Int!
  limit: Int!
  total: Int!
  items: [SectionBlockText]!
}

type Page implements Entry {
  title: String!
  subpages: PageCollection
  meta: PageMetadata
  sections: PageSectionCollection
  flags: [PageFlag]
}

type PageCollection {
  skip: Int!
  limit: Int!
  total: Int!
  items: [Page]!
}

type PageMetadata implements Entry {
  canonicalLink: String
  metaKeywords: String
  metaFlag: [PageMetadataMetaFlag]
  structuredData: JSON
}

enum PageMetadataMetaFlag {
  no-follow
  no-index
}

\\"\\"\\"
The \`JSON\` scalar type represents JSON values as specified by [ECMA-404](http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf).
\\"\\"\\"
scalar JSON @specifiedBy(url: \\"http://www.ecma-international.org/publications/files/ECMA-ST/ECMA-404.pdf\\")

type PageSectionCollection {
  skip: Int!
  limit: Int!
  total: Int!
  items: [PageSection]!
}

union PageSection = SectionBlockText | SectionCarousel

type SectionCarousel implements Entry {
  images: AssetCollection
  slideInterval: Float
}

enum PageFlag {
  Not Shareable
  Set Campus Cookie
}

type SectionCarouselCollection {
  skip: Int!
  limit: Int!
  total: Int!
  items: [SectionCarousel]!
}

type PageMetadataCollection {
  skip: Int!
  limit: Int!
  total: Int!
  items: [PageMetadata]!
}
"
`;
